lib/uklibparam/parser.c

```diff
@@ -0,0 +1,707 @@
+	 *       parse a number value for it.
+	 */
+	UK_ASSERT(ctx->p->type != _UK_LIBPARAM_PARAM_TYPE_charp);
+
+	/* From here on, we cannot continue without having a value to convert */
+	if (!value || (value_len == 0))
+		goto novalue;
+
+	/* Parse `value` and store result in `num` */
+	len = __safe_strntoull(value, value_len, __NULL,
+			       0 /* auto-detect base */,
+			       &num, &is_negative);
+	if ((len < 0) || ((__sz) len < value_len)) {
+		if (len == -ERANGE)
+			goto outofrange;
+
+		uk_pr_debug("len:%"__PRIssz" value_len:%"__PRIsz"\n",
+			    len, value_len);
+		goto malformed;
+	}
+
+	/* Check if resulting number fits */
+#define do_check_num(num, is_negative, min, max)			\
+	do {								\
+		if (num > (unsigned long long) max)			\
+			goto outofrange;				\
+		if (is_negative) {					\
+			num = -num;					\
+			if ((long long) num < (long long) min)		\
+				goto outofrange;			\
+		}							\
+	} while (0)
+
+	switch (ctx->p->type) {
+	case _UK_LIBPARAM_PARAM_TYPE___s8:
+		do_check_num(num, is_negative, __S8_MIN, __S8_MAX);
+		break;
+	case _UK_LIBPARAM_PARAM_TYPE___s16:
+		do_check_num(num, is_negative, __S16_MIN, __S16_MAX);
+		break;
+	case _UK_LIBPARAM_PARAM_TYPE___s32:
+		do_check_num(num, is_negative, __S32_MIN, __S32_MAX);
+		break;
+	case _UK_LIBPARAM_PARAM_TYPE___s64:
+		do_check_num(num, is_negative, __S64_MIN, __S64_MAX);
+		break;
+	case _UK_LIBPARAM_PARAM_TYPE___u8:
+		do_check_num(num, is_negative, 0, __U8_MAX);
+		break;
+	case _UK_LIBPARAM_PARAM_TYPE___u16:
+		do_check_num(num, is_negative, 0, __U16_MAX);
+		break;
+	case _UK_LIBPARAM_PARAM_TYPE___u32:
+		do_check_num(num, is_negative, 0, __U32_MAX);
+		break;
+	case _UK_LIBPARAM_PARAM_TYPE___u64:
+		do_check_num(num, is_negative, 0, __U64_MAX);
+		break;
+	case _UK_LIBPARAM_PARAM_TYPE___uptr:
+		do_check_num(num, is_negative, 0, __PTR_MAX);
+		break;
+	default:
+		/* No checks or modification needed for bool */
+		break;
+	}
+#undef do_check_num
+
+	/* Write number to target parameter */
+	return write_value(ctx->p, (__uptr) num);
+
+novalue:
+	uk_pr_err("Parameter %s.%s requires a %s value\n",
+		  ctx->ld->prefix, ctx->p->name, str_param_type(ctx->p->type));
+	return -EINVAL;
+
+outofrange:
+	uk_pr_err("Parameter %s.%s (%s): Given number is out of range\n",
+		  ctx->ld->prefix, ctx->p->name, str_param_type(ctx->p->type));
+	return -ERANGE;
+
+malformed:
+	uk_pr_err("Parameter %s.%s (%s): Given number is malformed\n",
+		  ctx->ld->prefix, ctx->p->name, str_param_type(ctx->p->type));
+	return -EINVAL;
+}
+
+static int parse_arg(struct parse_arg_ctx *ctx, char *strbuf, int scan_only)
+{
+	UK_ASSERT(ctx);
+	UK_ASSERT(strbuf);
+
+	char *value;
+	__sz value_len;
+	int ret = 0;
+
+	if (strbuf[0] == '\0') {
+		 /* Empty string: We skip this snippet */
+		return 0;
+	}
+
+	uk_pr_debug("Parsing snippet: \"%s\"\n", strbuf);
+	switch (ctx->state) {
+	case PAS_PARAM:
+		/*
+		 * Catch stop sequence ('---') or usage command ('help')
```

Comment:
```suggestion
		 * Catch stop sequence ('--') or usage command ('help')
```
Based on `#define PARSE_STOP         "--"`